src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "utils/concepts-enabled.h"
+#include "utils/gate_guard.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/internal/pollable_fd.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/smp.hh>
+
+#include <boost/lockfree/spsc_queue.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+#include <mutex>
+
+namespace v8_engine {
+
+namespace internal {
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() noexcept = 0;
+    virtual void cancel() noexcept = 0;
+    virtual void on_timeout() noexcept = 0;
+    virtual void set_exception(const std::exception_ptr& exception) {
+        _exception = exception;
+    };
+    virtual std::chrono::milliseconds get_timeout() const noexcept = 0;
+
+    virtual void done() = 0;
+
+    std::exception_ptr _exception;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    task(Func&& f, std::chrono::milliseconds timeout)
+      : work_item()
+      , _func(std::forward<Func>(f))
+      , _timeout(timeout) {}
+
+    // Save first exception for task
+    template<typename FunForExecute>
+    void func_wrapper(FunForExecute&& func) {
+        if (_exception) {
+            return;
+        }
+
+        try {
+            func();
+        } catch (...) {
+            set_exception(std::current_exception());
+        }
+    }
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() noexcept override {
+        func_wrapper([this] { _func(); });
+    }
+
+    // Run this function for cancel task execution
+    void cancel() noexcept override {
+        func_wrapper([this] { _func.cancel(); });
+    }
+
+    // Run this function after wathdog was alarmed
+    void on_timeout() noexcept override {
+        func_wrapper([this] { _func.on_timeout(); });
+    }
```

Comment:
Nit: it is possible to do this without lambdas using pointer to method non-type template parameter

```suggestion
    // Save first exception for task
    template <void (Func::*method)()>
    void func_wrapper() noexcept {
        if (_exception) {
            return;
        }
        try {
            (_func.*method)();
        } catch (...) {
            set_exception(std::current_exception());
        }
    }

    // Process task from executor in executor thread and set state in ss::future
    void process() noexcept override {
        func_wrapper<&Func::operator()>();
    }
    
    // Run this function for cancel task execution
    void cacnel() noexcept override {
        func_wrapper<&Func::cancel>();
    }
    
    // Run this function after wathdog was alarmed
    void on_timeout() noexcept override {
        func_wrapper<&Func::on_timeout>();
    }        
```

Reply:
done

Reply:
It does not work with v8_script executor task(

Reply:
this will not work as is with inheritance and virtual methods, but will work with non-virtual interface